server/service/service_campaigns.go

```diff
@@ -53,26 +54,35 @@ func (svc Service) NewDistributedQueryCampaign(ctx context.Context, queryString
 	}
 
 	var query *fleet.Query
+	var err error
 	if queryID != nil {
-		query, err := svc.ds.Query(*queryID)
+		query, err = svc.ds.Query(*queryID)
 		if err != nil {
 			return nil, err
 		}
 		queryString = query.Query
 	} else {
+		if err := svc.authz.Authorize(ctx, &fleet.Query{}, fleet.ActionWrite); err != nil {
+			return nil, err
+		}
+		err := query.ValidateSQL()
+		if err != nil {
+			return nil, err
+		}
 		query = &fleet.Query{
 			Name:     fmt.Sprintf("distributed_%s_%d", vc.Email(), time.Now().Unix()),
 			Query:    queryString,
 			Saved:    false,
 			AuthorID: ptr.Uint(vc.UserID()),
 		}
+		query, err = svc.ds.NewQuery(query)
+		if err != nil {
+			return nil, errors.Wrap(err, "new query")
+		}
 	}
-	if err := query.ValidateSQL(); err != nil {
-		return nil, err
-	}
-	query, err := svc.ds.NewQuery(query)
-	if err != nil {
-		return nil, errors.Wrap(err, "new query")
+
+	if null.StringFromPtr(vc.User.GlobalRole).ValueOrZero() == fleet.RoleObserver && !query.ObserverCanRun {
```

Comment:
I noticed that we are not checking the role and whether an observer can run a query before running it either.

Reply:
This should be handled by the authorization logic tested in https://github.com/fleetdm/fleet/blob/main/server/authz/policy_test.go#L359-L366 and implemented in https://github.com/fleetdm/fleet/blob/main/server/authz/policy.rego#L277-L292.

Are we missing a call to the authorization check in this method though?

If we are indeed missing that, I'm not sure how any live queries are running successfully due to the middleware that fails any request that has not been authorized (or explicitly skipped). So maybe I'm missing something.

Reply:
I didn't see any calls to Authorize except the one I added. I'll add a run check instead of this.

Reply:
All right, should be addressed here: 3de8496